src/main/kotlin/com/zepben/evolve/cim/iec61970/base/core/Curve.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 Zeppelin Bend Pty Ltd
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+package com.zepben.evolve.cim.iec61970.base.core
+
+import com.zepben.evolve.services.common.extensions.asUnmodifiable
+import com.zepben.evolve.services.common.extensions.typeNameAndMRID
+
+
+/**
+ * The Curve class is a multipurpose functional relationship between an independent variable (X-axis) and dependent (Y-axis) variables.
+ */
+abstract class Curve @JvmOverloads constructor(mRID: String = "") : IdentifiedObject(mRID) {
+
+    private var _data: MutableList<CurveData>? = null
+
+    /**
+     * Add a data point to this [Curve].
+     *
+     * @param x The data value of the X-axis variable, depending on the X-axis units.
+     * @param y1 The data value of the first Y-axis variable, depending on the Y-axis units.
+     * @param y2 The data value of the second Y-axis variable (if present), depending on the Y-axis units.
+     * @param y3 The data value of the third Y-axis variable (if present), depending on the Y-axis units.
+     * @throws IllegalArgumentException if a [CurveData] for the provided [x] value already exists for this Curve.
+     */
+    fun addData(x: Float, y1: Float, y2: Float? = null, y3: Float? = null): Curve {
+        require(_data.isNullOrEmpty() || _data?.none { cd -> cd.xValue == x } == true) {
+            "Unable to add datapoint to ${typeNameAndMRID()}. " +
+                "xValue $x is invalid. as data with same xValue already exist in this Curve. "
+        }
+        _data = _data.or(::mutableListOf) {
+            add(CurveData(x, y1, y2, y3))
+            sortBy { it.xValue }
+        }
+        return this
+    }
+
+    /**
+     * Add a data point to this [Curve].
+     */
+    fun addData(curveData: CurveData): Curve = addData(curveData.xValue, curveData.y1Value, curveData.y2Value, curveData.y3Value)
+
+    /**
+     * The point data values that define this curve. The returned collection is read only.
+     * The curveData in the underlying collection will be sorted by xValue in ascending order.
+     */
+    val data: List<CurveData> get() = _data.asUnmodifiable()
+
+    /**
+     * Get point data values by its xValue.
+     */
+    fun getData(x: Float): CurveData? = _data?.find { it.xValue == x }
+
+    /**
+     * A curve object can have 0 or more curveData, index in the list indicates nth x value of stored curveData.
+     */
+    private fun getData(sequenceNumber: Int): CurveData? = _data?.get(sequenceNumber)
+
+    /**
+     * Get point data values by its sequenceNumber.
```

Comment:
todo: Missing @param in docstring

Reply:
changed to use xValue instead of sequenceNumber and have the same docstring as the getData function.